Analysis decorator fixes #946

kain88-de opened the pull request:
Fixes #946

Changes made in this Pull Request:
 - Added `AnalysisFromFunction` class
```python
 >>> def rotation_matrix(mobile, ref):
 >>>     return mda.analysis.align.rotation_matrix(mobile, ref)[0]
 >>> rot = AnalysisFromFunction(rotation_matrix, trajectory, mobile, ref).run()
 >>> print(rot.results)
```
 - Added `analysis_class` decorator
```python
>>> @analysis_class
>>> def RotationMatrix(mobile, ref):
>>>     return mda.analysis.align.rotation_matrix(mobile, ref)[0]
>>>
>>> rot = RotationMatrix(u.trajectory, mobile, ref, step=2).run()
>>> print(rot.results)
```

Trajectory is a argument that is added now and required. I guess it should be possible to infer it from the first AtomGroup found in `*args` or `**kwargs` if it is not specified.

PR Checklist
------------
 - [x] Tests?
 - [x] Docs?
 - [x] CHANGELOG updated?
 - [x] Issue raised/referenced?
 - [x] infer trajectory from passed arguments

kain88-de commented:
@jbarnoud are the two options to your liking? I went with a dedicated class instead of `RotMatrix = analysis_class(rotation_matrix)` that you suggested as I feel this makes it easier for people who are new to python.

richardjgowers commented on package/MDAnalysis/analysis/base.py:
```diff
@@ -153,8 +159,8 @@ def run(self):
         self._prepare()
-        for i, ts in enumerate(
-                self._trajectory[self.start:self.stop:self.step]):
+        for i, ts in enumerate(self._trajectory[self.start:self.stop:
```
This looks like a pretty ugly change?

richardjgowers commented on testsuite/MDAnalysisTests/analysis/test_base.py:
```diff
@@ -93,3 +93,28 @@ def test_incomplete_defined_analysis(self):
+
+        kwargs = {'step': 3, 'foo': None}
+
+        with raises(ValueError):
```
This should be split out into its own test

jbarnoud commented on package/MDAnalysis/analysis/base.py:
```diff
@@ -22,17 +21,19 @@
 """
-from six.moves import range
-
+import inspect
```
Compatibility imports should come first as they can change the language behavior.

jbarnoud commented on package/MDAnalysis/analysis/base.py:
```diff
@@ -163,3 +169,112 @@ def run(self):
+    base_argspec = inspect.getargspec(AnalysisBase.__init__)
+    n_base_defaults = len(base_argspec.defaults)
+    base_kwargs = {name: val
+                    for name, val in zip(base_argspec.args[
```
ugly line split, maybe you could build the zip on a separate line.

jbarnoud commented:
@kain88-de Great! Yet, I would probably use the `analysis_class` as a function to produce a class rather than a decorator, or the `AnalysisFromFunction` class directly. Maybe we have a slightly different view of the thing. As I see it, I want to write a new analysis tool, so I write the core function, I create the class from the core function, and the user uses it as if I wrote the full fledge class. As I see it, the end-user does not have to know that a given analysis was written with all the boilerplate, or with the shortcut. From what you suggest, the decorator does what I expect from the feature, but is confusing when reading the code. Using directly the class is fine for tweaking stuff, but the end-user needs to know that this analysis in particular is a function and not a class.

My view is perfectly fulfilled with `analysis_class` used as a function. So the discrepancies between our point of views is only in the usage, good practices, and therefore doc.

jbarnoud commented on testsuite/MDAnalysisTests/analysis/test_base.py:
```diff
@@ -93,3 +93,28 @@ def test_incomplete_defined_analysis(self):
+
+
+class TestAnalysisFunction(object):
+    def test_filter_kwargs():
```
`self` is missing (or `@staticmethod` to make code quantify happy)

jbarnoud commented on package/MDAnalysis/analysis/base.py:
```diff
@@ -163,3 +169,112 @@ def run(self):
         return self
+
+
+def _filter_kwargs(function, kwargs):
```
According to PEP8, functions should go after the classes.

jbarnoud commented:
`AnalysisFronFunction.function` could be a class attribute rather than an argument of `__init__`. Then, `analysis_class` would be

```python
def analysis_class(function):
    class WrapperClass(AnalysisFromFunction):
        function=function
    return WrapperClass
```

This minor change would allow to inherit from the newly created class, or to monkey patch additional methods.

kain88-de commented:
> @kain88-de Great! Yet, I would probably use the analysis_class as a function to produce a class rather than as a decorator, or the AnalysisFromFunction class directly. Maybe we have a slightly different view of the feature. As I see it, I want to write a new analysis tool, so I write the core function, I create the class from the core function, and the user uses it as if I wrote the full fledge class. 

OK what I though `AnalysisFromFunction` would be used from a user with his own functions, (handy in interactive session).
`analysis_class` is a decorator that can be used by a developer or a user in both syntax ways that you laid out. But I haven't checked yet how the docs for the decorated functions are looking and if they make sense.

> As I see it, the end-user does not have to know that a given analysis was written with all the boilerplate, or with the shortcut. From what you suggest, the decorator does what I expect from the feature, but is confusing when reading the code. Using directly the class is fine for tweaking stuff, but the end-user needs to know that this analysis in particular is a function and not a class.

You can use the `analysis_class` function in both ways. As long as we document it is should be fine.


> This minor change would allow to inherit from the newly created class, or to monkey patch additional methods.

Why would we want to do that? I personally don't like monkey-patching because it makes it hard to see from the code what it will do at runtime. I explained above that I think `AnalysisFromFunction` is a handy class for a user in itself so I prefer it like this.

jbarnoud commented:
The monkey patching is just a example, but the inheritance can come handy. One thing I like with the analysis class is that you can define properties to give access to different views of the results. In your RotationMatrix example, one could add a `as_radiant` properties by doing:

```python
def rotation_matrix(mobile, ref):
    return mda.analysis.align.rotation_matrix(mobile, ref)[0]

RotationMatrix = analysis_class(rotation_matrix)

class ImprovedRotationMatrix(RotationMatrix):
    @property
    def as_radiant(self):
        return 2 * numpy.pi * self.result / 360
```

With monkey patching it would not be a property, but it would look like:

```python
def rotation_matrix(mobile, ref):
    return mda.analysis.align.rotation_matrix(mobile, ref)[0]

def as_radiant(self):
     return 2 * numpy.pi * self.result / 360

RotationMatrix = analysis_class(rotation_matrix)
RotationMatrix.as_radiant = as_radiant
```

One or the other, they are compact and easy ways to add features to analysis classes built with the `analysis_class` shortcut. Of course, at some point, you will have to write a full fledge class if you want to add more and more features, but for simple things it is a simple approach.

> OK what I though AnalysisFromFunction would be used from a user with his own functions, (handy in interactive session).

I would say that in an interactive session, you want your initial function to run untouched to make sure it works. Therefore, especially in an interactive session, you better not use `analysis_class` as a decorator. But this is only my work flow.

My take on it would be to show in the doc the two ways of doing it (as a decorator, and as a function), but to encourage the use as a function for the reasons I gave above (less confusing, you still can use your function).

> But I haven't checked yet how the docs for the decorated functions are looking and if they make sense.

This is a good point. We may have to copy the function doc in the new class. But it is likely not to make sense. Yet, getting a class would allow to patch the doc:

```python
def rotation_matrix(mobile, ref):
    return mda.analysis.align.rotation_matrix(mobile, ref)[0]

RotationMatrix = analysis_class(rotation_matrix)
RotationMatrix.__doc__ = """
The awesome docstring of my great analysis class

Parameters
==========
trajectory...
"""
```

kain88-de commented:
I see that inheriting might be since. I see what I can do. But I would really like that is still possible to use `AnalysisFromFunction` still as described above.

jbarnoud commented:
```python
def analysis_class(function):
    class WrapperClass(AnalysisFromFunction):
       function = function
        def __init__(self, trajectory, *args, **kwargs):
            super(WrapperClass, self).__init__(self, self.function, trajectory, *args, **kwargs)

    return WrapperClass
```

Maybe?

kain88-de commented:
Addressed all comments. I've added extensive tests. I welcome suggestions for more tests as I might have missed some edge-case. I also welcome suggested alternatives for `AnalysisFromFunction`, I simply choose it because I couldn't think of any better.

jbarnoud commented on package/MDAnalysis/analysis/base.py:
```diff
@@ -163,3 +167,151 @@ def run(self):
+
+
+def analysis_class(function):
+    """decorator to transform a function operating on a single frame to an
```
I would formulate it as

```
Transform a function operating on a single frame to an analysis class

For an usage in a library we recommend the following style:

>>> def simple(mobile):
>>>     pass
>>> SimpleAnalysis = analysis_class(simple)

It can also be used as a decorator:

>>> @analysis_class
 >>> def RotationMatrix(mobile, ref):
>>>     return mda.analysis.align.rotation_matrix(mobile, ref)[0]

>>> rot = RotationMatrix(u.trajectory, mobile, ref, step=2).run()
>>> print(rot.results)
```

Then of course, the example would go away.

The point is to have both example together rather than scattered throughout the docstring. Also, the change in the initial description remove the emphasis on the decorator aspect to match with the recommendation just bellow.

jbarnoud commented on package/MDAnalysis/analysis/base.py:
```diff
@@ -163,3 +167,151 @@ def run(self):
+                for arg in six.itervalues(kwargs):
+                    if isinstance(arg, AtomGroup):
+                        trajectory = arg.universe.trajectory
+
```
At that point it is still possible that *trajectory* is `None` and that none of the argument was an `AtomGroup`. Maybe this should trigger an error.

kain88-de replied:
It would trigger an error in AnalysisBase

jbarnoud replied:
Super.

kain88-de replied:
I made it explicit now. Should be better then some deep stack that has to be inspected.

jbarnoud commented on testsuite/MDAnalysisTests/analysis/test_base.py:
```diff
@@ -93,3 +94,64 @@ def test_incomplete_defined_analysis(self):
+    ana = base.AnalysisFromFunction(simple_function, mobile=u.atoms,
+                                    step=step).run()
+
+    ana = base.AnalysisFromFunction(simple_function, u.atoms,
```
What are these two redundant calls for?

kain88-de replied:
Testing different options in the arguments.

jbarnoud replied:
Then, maybe should we make sure that they give the same result. Maybe they could be saved in `ana1` and `ana2`, which could both be checked against the reference.

kain88-de replied:
done

jbarnoud commented:
I will play a bit with it tonight or tomorrow. Then, if nobody objects, I'll merge.
